File: 25/go/validator/validator.go
package validator

import (
	"unicode"
)

// Strength 強さ
type Strength int

const (
	// VeryWeak 最弱
	VeryWeak Strength = iota
	// Weak 弱
	Weak
	// Normal ふつう
	Normal
	// Strong 強
	Strong
	// VeryStrong 最強
	VeryStrong
)

func (o Strength) String() string {
	switch o {
	case VeryWeak:
		return "very weak"
	case Weak:
		return "weak"
	case Normal:
		return "normal"
	case Strong:
		return "strong"
	case VeryStrong:
		return "very strong"
	default:
		break
	}
	panic("must not reach.")
}

func isAlpha(r rune) bool {
	if ('a' <= r && r <= 'z') || ('A' <= r && r <= 'Z') {
		return true
	}
	return false
}

const passwordLenghtMin = 8

// PasswordValidator パスワード強度
func PasswordValidator(password string) Strength {
	numberOnly := allOf(password, unicode.IsDigit)
	letterOnly := allOf(password, isAlpha)
	containsSpecial := anyOf(password, unicode.IsPunct)
	containsLetter := anyOf(password, isAlpha)
	containsNumber := anyOf(password, unicode.IsDigit)
	tooShort := len(password) < passwordLenghtMin

	if tooShort {
		if numberOnly {
			return VeryWeak
		} else if letterOnly {
			return Weak
		}
	} else {
		if containsLetter && containsNumber && containsSpecial {
			return VeryStrong
		}
		if containsLetter && containsNumber {
			return Strong
		}
	}
	return Normal
}
func allOf(s string, f func(rune) bool) bool {
	for _, r := range s {
		if !f(r) {
			return false
		}
	}
	return true
}
func anyOf(s string, f func(rune) bool) bool {
	for _, r := range s {
		if f(r) {
			return true
		}
	}
	return false
}

File: 25/python/validator.py
#!/usr/bin/env python
# coding: -utf8

import enum
import string


PASSWORD_LENGHT_MIN = 8

Strength = enum.Enum("Strength", "VERY_WEAK  WEAK NORMAL STRONG VERY_STRONG")


def any_of(a, f):
    for c in a:
        if f(c):
            return True
    return False


def is_punct(s):
    return s in string.punctuation


def is_alpha(s):
    return s.isalpha()


def is_digit(s):
    return s.isdigit()


def password_validator(password):
    number_only = password.isdigit()
    letter_only = password.isalpha()
    contains_special = any_of(password, is_punct)
    contains_letter = any_of(password, is_alpha)
    contains_number = any_of(password, is_digit)
    too_short = len(password) < PASSWORD_LENGHT_MIN

    if too_short:
        if number_only:
            return Strength.VERY_WEAK
        elif letter_only:
            return Strength.WEAK
    else:
        if contains_letter and contains_number and contains_special:
            return Strength.VERY_STRONG
        if contains_letter and contains_number:
            return Strength.STRONG

    return Strength.NORMAL
